### PR TITLE
fix(android): handle `IllegalArgumentException` when initializing `CloudDownloadMgr`, add logging to check for unhandled side-effects

### PR DESCRIPTION
**Pull request by jahorton**
Replaces #11580.
Addresses #11561.  This change should prevent crashes, but there may be lingering side-effects we want to resolve before closing the issue outright.

This PR does two main things:
1. It ensures we `catch` the _right_ exception type - `IllegalArgumentException` instead of `IllegalStateException` - in our existing error handlers.  This should, at least, prevent a crash.
2. We didn't have manual breadcrumb logging yet within Keyman Engine for Android, so this adds a method to accomplish that.

Toward that latter point, here's how a stack-traced breadcrumb appears in Sentry:

image

Here's a link to the Sentry event I temporarily generated in order to expose this: https://keyman.sentry.io/share/issue/0d3661c2239842ff8ba31bed996087ec/

The `stacktrace` entry is initially collapsed, but expands when clicked.  To keep the message-envelope size down, breadcrumb-logging will only include the 10 most recent entries of the stack trace.

@keymanapp-test-bot skip

### CODE REVIEW

File: android/KMEA/app/src/main/java/com/keyman/engine/util/KMLog.java [hunk 1]
```diff
@@ -34,6 +35,44 @@ public static void LogInfo(String tag, String msg) {
...
+        Breadcrumb crumb = new Breadcrumb();
+        crumb.setMessage(msg);
+        crumb.setLevel(SentryLevel.INFO);
+
+        if(addStackTrace) {
+          StackTraceElement[] rawTrace = Thread.currentThread().getStackTrace();
+
+          // The call that gets us the stack-trace above... shows up in the
+          // stack trace, so we'll skip the first few (redundant) entries.
+          int skipCount = 3;
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/keyman/engine/util/KMLog.java:58]: Stack-trace generation produces the same trace as is viewable for Sentry exceptions, etc.  So, compare against the snippet below:

image

The breadcrumb-logging method is entry 3 within the stack trace, so we skip the first three entries.  Granted, it's _possible_ the count might differ a bit when on non-emulated devices, but even then, we should only skip the parts that we know make the message.

**Comment by mcdurdin** [comment]
> feat(android): cloud-downloader state logging

is this a fix or a feature? Or both? Seems like the crash fix should be in the title. (Note, title is also not imperative)

**Comment by mcdurdin** [comment]
Suggested title:

> fix(android): handle `IllagalArgumentException` when initializing `CloudDownloadMgr` and add logging

**Comment by jahorton** [comment]
> is this a fix or a feature? Or both? Seems like the crash fix should be in the title. (Note, title is also not imperative)

"feat" (or, "feature") is not treatable as an imperative verb?  Granted, "fix" is certainly more straightforward in that regard.

I hesitate to call it a fix.  It does patch up an error-handler, but we don't think the issue is fully-fixed yet, do we?

**Comment by mcdurdin** [comment]
> "feat" (or, "feature") is not treatable as an imperative verb?

We should still use imperative mood for the title. The type is not a part of the title sentence (which is partly why I am not too hung up about repetition `fix(scope): fix the thing`; all the types are nouns (even if they can also be verbs) -- 'fix', 'chore', 'docs', 'change', etc.

> I hesitate to call it a fix. It does patch up an error-handler, but we don't think the issue is fully-fixed yet, do we?

It's still a partial fix, and we expect it will mitigate the issue. A fix doesn't have to be 100% to be valuable. But it's certainly not a feature -- at best logging is a chore.

File: android/KMEA/app/src/main/java/com/keyman/engine/util/KMLog.java [hunk 2]
```diff
@@ -34,6 +35,44 @@ public static void LogInfo(String tag, String msg) {
...
+   * Utility to log info and add as a Sentry breadcrumb, rather than
+   * as an independent message
+   * @param tag String of the caller
+   * @param msg String of the info message
+   */
+  public static void LogBreadcrumb(String tag, String msg, boolean addStackTrace) {
+    if (msg != null && !msg.isEmpty()) {
+      Log.i(tag, msg);
+
+      if (DependencyUtil.libraryExists(LibraryType.SENTRY) && Sentry.isEnabled()) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/keyman/engine/util/KMLog.java:48]: nit, but it's better for preconditions to exit early to reduce brace depth:

```suggestion
    if (msg == null || msg.isEmpty()) {
      return;
    }
    
    Log.i(tag, msg);

    if (!DependencyUtil.libraryExists(LibraryType.SENTRY) || !Sentry.isEnabled()) {
      return;
    }
```

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.50-alpha](https://keyman.com/downloads/releases/alpha/18.0.50)

**Pull request merged**